src/repr/src/row/collection.rs

```diff
@@ -0,0 +1,731 @@
+    pub fn new(rows: Vec<(Row, NonZeroUsize)>) -> Self {
+        // Pre-sizing our buffer should allow us to make just 1 allocation, and
+        // use the perfect amount of memory.
+        //
+        // Note(parkmycar): I didn't do any benchmarking to determine if this
+        // is faster, so feel free to change this if you'd like.
+        let encoded_size = rows.iter().map(|(row, _diff)| row.data.len()).sum();
+
+        let mut encoded = Vec::<u8>::with_capacity(encoded_size);
+        let mut metadata = Vec::<EncodedRowMetadata>::with_capacity(rows.len());
+
+        for (row, diff) in rows {
+            encoded.extend(row.data());
+            metadata.push(EncodedRowMetadata {
+                offset: encoded.len(),
+                diff,
+            });
+        }
+
+        RowCollection {
+            encoded: Bytes::from(encoded),
+            metadata,
+        }
+    }
+
+    /// Merge another [`RowCollection`] into `self`.
+    pub fn merge(&mut self, other: &RowCollection) {
+        if other.count() == 0 {
+            return;
+        }
+
+        // TODO(parkmycar): Using SegmentedBytes here would be nice.
+        let mut new_bytes = vec![0; self.encoded.len() + other.encoded.len()];
+        new_bytes[..self.encoded.len()].copy_from_slice(&self.encoded[..]);
+        new_bytes[self.encoded.len()..].copy_from_slice(&other.encoded[..]);
+
+        let mapped_metas = other.metadata.iter().map(|meta| EncodedRowMetadata {
+            offset: meta.offset + self.encoded.len(),
+            diff: meta.diff,
+        });
+
+        self.metadata.extend(mapped_metas);
+        self.encoded = Bytes::from(new_bytes);
+    }
+
+    /// Total count of [`Row`]s represented by this collection.
+    pub fn count(&self) -> usize {
+        self.metadata.iter().map(|meta| meta.diff.get()).sum()
+    }
+
+    /// Total count of ([`Row`], [`EncodedRowMetadata`]) pairs in this collection.
+    pub fn entries(&self) -> usize {
+        self.metadata.len()
+    }
+
+    /// Returns the number of bytes this [`RowCollection`] uses.
+    pub fn byte_len(&self) -> usize {
+        let row_data_size = self.encoded.len();
+        let metadata_size = self
+            .metadata
+            .len()
+            .saturating_mul(std::mem::size_of::<EncodedRowMetadata>());
+
+        row_data_size.saturating_add(metadata_size)
+    }
+
+    /// Returns a [`RowRef`] for the entry at `idx`, if one exists.
+    pub fn get(&self, idx: usize) -> Option<(&RowRef, &EncodedRowMetadata)> {
+        let (lower_offset, upper) = match idx {
+            0 => (0, self.metadata.get(idx)?),
+            _ => {
+                let lower = self.metadata.get(idx - 1).map(|m| m.offset)?;
+                let upper = self.metadata.get(idx)?;
+                (lower, upper)
+            }
+        };
+
+        let slice = &self.encoded[lower_offset..upper.offset];
+        // SAFETY: We know all of the data in a RowCollection are valid Rows.
```

Comment:
That's a great callout! Without using `ProtoRow`, which we want to avoid, I'm not sure how we could _guarantee_ the bytes are valid Row data? Theoretically our tests should catch if someone inserted non-valid data though?

Do reduce the risk maybe we don't implement `Deserialize`?

Reply:
Yeah, I think the only way to guarantee that is to parse every row during deserialization, and error if parsing fails, which totally defeats the purpose of doing this optimization I think.

How bad is it if `RowRef` points to invalid data? Would it lead to undefined behavior or could we make methods on it panic instead, when they detect invalid data? That would let us at least rule out UB.

> Do reduce the risk maybe we don't implement `Deserialize`?

If we don't need it, that sounds like a good idea!

Reply:
Chatted with @teskje offline, if `RowRef` contains invalid data it should panic on read. I removed the `Deserialize` impl because it wasn't needed and reduced our risk surface here.

Otherwise there isn't a way to _guarantee_ our Row data is valid, but a dev would have to manually create a `ProtoRowCollection`, which nothing does currently, to expose an issue.